867 simplify access control structure

alvacouch opened the pull request:
Issue #867: improvements to access control
* Eliminate one join from all queries. 
    * user__uaccess__u2xxx --> user__r2xxx for all "xxx"
    * group__gacess__g2xxx --> group__g2xxx
    * resource__raccess__r2xxx -> resource__r2xxx 
* Raise standard exception PermissionDenied instead of custom exceptions. Eliminate custom exceptions. 
* Enhance anti-bugging assertions during debug mode; eliminate from production mode. 
* Complete implementation of User.is_active flag: 
    * inactive users disappear from groups, shares, and other user lists. 
    * acting as or on an inactive user raises PermissionDenied 
* Correct implementation of share_Y_with_Y: 
    * undo now prohibits unshare with self. 
    * unshare_X_with_Y matches can_unshare_X_with_Y

hydrobuild commented:
Can one of the admins verify this patch?

mjstealey commented:
@alvacouch - Though this passes all tests will still need to look at the migrations using the existing www database.  Did a test deploy on [https://alpha.hydroshare.org](https://alpha.hydroshare.org) which didn't want to accept all migrations files. End result was users not having any resources associated with them.

Example console output:
```bash
Running migrations:
  Applying hs_access_control.0006_auto_add_new_fields... OK
  Applying hs_access_control.0007_manual_populate_new_fields... OK
  Applying hs_access_control.0008_auto_remove_many2many_relationships... OK
  Applying hs_access_control.0009_auto_remove_original_fields... OK
  Applying hs_access_control.0010_auto_rename_related_names... OK
  Applying hs_access_control.0011_auto_rename_new_fields_to_original_names... OK
  Applying hs_access_control.0012_auto_disallow_nulls...Traceback (most recent call last):
  File "manage.py", line 8, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/__init__.py", line 385, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/__init__.py", line 377, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/base.py", line 288, in run_from_argv
    self.execute(*args, **options.__dict__)
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/base.py", line 338, in execute
    output = self.handle(*args, **options)
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/commands/migrate.py", line 161, in handle
    executor.migrate(targets, plan, fake=options.get("fake", False))
  File "/usr/local/lib/python2.7/dist-packages/django/db/migrations/executor.py", line 68, in migrate
    self.apply_migration(migration, fake=fake)
  File "/usr/local/lib/python2.7/dist-packages/django/db/migrations/executor.py", line 102, in apply_migration
    migration.apply(project_state, schema_editor)
  File "/usr/local/lib/python2.7/dist-packages/django/db/migrations/migration.py", line 108, in apply
    operation.database_forwards(self.app_label, schema_editor, project_state, new_state)
  File "/usr/local/lib/python2.7/dist-packages/django/db/migrations/operations/fields.py", line 139, in database_forwards
    schema_editor.alter_field(from_model, from_field, to_field)
  File "/usr/local/lib/python2.7/dist-packages/django/db/backends/schema.py", line 470, in alter_field
    self._alter_field(model, old_field, new_field, old_type, new_type, old_db_params, new_db_params, strict)
  File "/usr/local/lib/python2.7/dist-packages/django/db/backends/schema.py", line 620, in _alter_field
    params,
  File "/usr/local/lib/python2.7/dist-packages/django/db/backends/schema.py", line 111, in execute
    cursor.execute(sql, params)
  File "/usr/local/lib/python2.7/dist-packages/django/db/backends/utils.py", line 81, in execute
    return super(CursorDebugWrapper, self).execute(sql, params)
  File "/usr/local/lib/python2.7/dist-packages/django/db/backends/utils.py", line 65, in execute
    return self.cursor.execute(sql, params)
  File "/usr/local/lib/python2.7/dist-packages/django/db/utils.py", line 94, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/usr/local/lib/python2.7/dist-packages/django/db/backends/utils.py", line 65, in execute
    return self.cursor.execute(sql, params)
django.db.utils.IntegrityError: column "grantor_id" contains null values
```

Full deployment console log is here: http://ci.hydroshare.org:8080/view/2.%20Deployment/job/branch-builder-with-irods/16/console

pkdash commented:
Due to large number of changes in models.py, github is not showing those changes. @mjstealey, @ironfroggy Is there way to see changes in models.py?

alvacouch commented:
See issue #889. This affects this pull and needs to be resolved first.

alvacouch commented:
The only way I can think of seeing changes in models.py is to
a) rebase the build so that the changes are staged. 
b) look at each wave of changes separately. 
This would take a new repository and PR. 

There is another way in Linux itself: 
diff -C 10 old-models.py new-models.py
will do what GitHub does and list changes in a similar format.

alvacouch commented:
@mjstealey: in migration 0007 I forgot to save() records. This is the cause of the above. Fixed in latest push.

ironfroggy commented:
You can easily very the diff in git without any rebasing or copying of the files between branches.

```
git fetch
git diff origin/develop origin/867-simplify-access-control-structure -- path/to/file
```

where `path/to/file` is the path in the repository to a specific file you want to see the diff between two branches for.

mjstealey commented:
> in migration 0007 I forgot to save() records. This is the cause of the above. Fixed in latest push.

@alvacouch - all of your migrations are now passing.
Branch has been re-deployed to https://alpha.hydroshare.org using www resources for testing.

A similar migration issue has been found in `hs_model_program` and a separate [issue](https://github.com/hydroshare/hydroshare/issues/892) has been created for it.
```bash
Operations to perform:
  Synchronize unmigrated apps: debug_toolbar, galleries, irods_browser_app, accounts, conf, rest_framework, forms, corsheaders, generic, boot, comments, blog, jquery_ui, haystack, crispy_forms, core, hs_metrics, filebrowser_safe, sitemaps, gis, pages, mobile, autocomplete_light, djcelery, widget_tweaks, django_extensions
  Apply all migrations: redirects, theme, hs_geo_raster_resource, hs_geographic_feature_resource, ga_resources, ga_ows, hs_labels, hs_model_program, sites, contenttypes, hs_tools_resource, hs_access_control, hs_app_netCDF, hs_modelinstance, ref_ts, sessions, hs_swat_modelinstance, auth, django_docker_processes, hs_app_timeseries, oauth2_provider, admin, hs_core, django_irods
Synchronizing apps without migrations:
  Creating tables...
  Installing custom SQL...
  Installing indexes...
Running migrations:
  Applying hs_access_control.0006_auto_add_new_fields... OK
  Applying hs_access_control.0007_manual_populate_new_fields... OK
  Applying hs_access_control.0008_auto_remove_many2many_relationships... OK
  Applying hs_access_control.0009_auto_remove_original_fields... OK
  Applying hs_access_control.0010_auto_rename_related_names... OK
  Applying hs_access_control.0011_auto_rename_new_fields_to_original_names... OK
  Applying hs_access_control.0012_auto_disallow_nulls... OK
  Applying hs_access_control.0013_auto_add_uniqueness_constraints... OK
  Applying hs_model_program.0006_auto_20151216_1500...Traceback (most recent call last):
  File "manage.py", line 8, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/__init__.py", line 385, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/__init__.py", line 377, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/base.py", line 288, in run_from_argv
    self.execute(*args, **options.__dict__)
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/base.py", line 338, in execute
    output = self.handle(*args, **options)
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/commands/migrate.py", line 161, in handle
    executor.migrate(targets, plan, fake=options.get("fake", False))
  File "/usr/local/lib/python2.7/dist-packages/django/db/migrations/executor.py", line 68, in migrate
    self.apply_migration(migration, fake=fake)
  File "/usr/local/lib/python2.7/dist-packages/django/db/migrations/executor.py", line 102, in apply_migration
    migration.apply(project_state, schema_editor)
  File "/usr/local/lib/python2.7/dist-packages/django/db/migrations/migration.py", line 108, in apply
    operation.database_forwards(self.app_label, schema_editor, project_state, new_state)
  File "/usr/local/lib/python2.7/dist-packages/django/db/migrations/operations/models.py", line 282, in database_forwards
    getattr(new_model._meta, self.option_name, set()),
  File "/usr/local/lib/python2.7/dist-packages/django/db/backends/schema.py", line 310, in alter_unique_together
    ", ".join(columns),
ValueError: Found wrong number (0) of constraints for hs_model_program_mpmetadata(content_type_id, object_id)
```

pkdash commented:
Since it is not possible to see the changes of models.py in github, I am going to provide comments in reference to line number.
Line#:201: Indentation (PEP8)

pkdash commented:
Line#:254: Unused variable ```access_group```
Line#:300: Is there a possibility this would return duplicate records (Group objects)?. The previous code used the ```distinct()```.
Line#:330 to 335: can be simplified replacing ```if``` with ```return```
Line#:362 to 367: can be simplified replacing ```if``` with ```return```

pkdash commented:
file: "test_access_control.py
Line#770 to 772: unused variables
Line#990, and 1014: should delete commented code
Line#3834 to 3840: should delete commented code
Line#4154 to 4234: unused variable ```foo```

pkdash commented on hs_core/views/__init__.py:
```diff
@@ -643,11 +645,13 @@ def _unshare_resource_with_users(request, requesting_user, users_to_unshare_with
                     requesting_user.uaccess.unshare_resource_with_user(resource, user)
                 else:
                     # requesting user is the original grantor of privilege to user
+		    # COUCH: This can raise a PermissionDenied exception without a guard such as 
```
PEP8 indentation

alvacouch replied:
Fixed. Was editing in VIM and it replaced spaces with tabs.

alvacouch commented:
@pkdash All review comments addressed, except: 

1. immutable flag is ready for a PR but not in this PR. Could include it now that base review has been completed. 
2. I need comments in "return False" to maintain the correspondences between "can_X" and "X" for methods X. 
3. distinct() is not needed in simple queries. It is needed in queries involving Q() expressions and |, because if both Q(A) and Q(B) match and we search for Q(A) | Q(B), we are guaranteed to have duplicates. My original version used distinct() too much.

alvacouch commented:
@pkdash: also removed some erroneously inserted tabs in various files... due to use of vim editor.

alvacouch commented:
@pkdash: Also, group semantics have not changed; this is a major change and is deferred until issue #893. This change is not ready to push.  So, at present, multiple people can indeed post the same privilege for the same group. This will be prevented in #893.

pkdash commented:
@alvacouch I am only seeing some of my comments being addressed. For example, original Line#1402 (new code line#1571): ```else``` is not needed. There are still those unused variables.

pkdash commented:
```get_groups_with_explicit_access()``` just a wrapper function for ```get_groups_with_declared_privilege()```. Why do we need the wrapper function?

pkdash commented:
modesl.py

Line#1670: why should we be raising exception if the resource has more than one owner?

pkdash commented:
May be it is time to split the test file. It is getting quite long.

alvacouch commented:
@pkdash 

Planning on splitting up tests in next version. Meanwhile, trying to get out a version that has full immutable semantics. 

Some responses: 

Line 1670: should not be able to remove last owner of resource. This exception reports that attempt. Otherwise, an owner would be able to remove the last owner (self). 

In general, in cases where I had
    if query.exists(): 
        return True
    else: 
       return False

This is not equivalent to 
    return query.exists()

The former returns bool. The latter returns unevaluated queryset. This is confusing. 

Will split test file next time around. There is a natural split between unit tests, functional tests, and integration tests. 

get_groups_with_combined_privilege has been eliminated from next version (being tested).

id__in and pk__in are equivalent. id__in is syntactic; pk__in is semantic and preferred. Changed all in next version to pk__in (under test).

pkdash commented:
@alvacouch Interesting that you are saying ```return query.exists()``` would not return a bool. I think have seen and even used myself in our codebase expecting it to return bool. This link https://docs.djangoproject.com/en/1.7/ref/models/querysets/#django.db.models.query.QuerySet.exists also says that it returns bool.  

I am OK with your current implementation as this is I guess just a style issue. If you are not planning to make any further changes for me to review then let me know so that I can +1 this PR.

alvacouch commented:
I am still working on this PR. I will post a new version with full immutable handling as soon as possible. I am still testing it. 

(It is the second day of classes and I am up to my ears in students who think their problems are more important than mine.)

alvacouch commented:
@pkdash This is likely to be the last change in this PR. Deferring other changes to later PRs. 

1. Eliminated all unused variables; (I hope) 
2. broke tests up into individual files with a helper library; 
3 changed "if p.exists(): return True" to "return p.exists()".
4. surveyed functions in use. 
5. standardized naming conventions. 
6. modified uses of methods system wide for new names. 
7. eliminated unused functions that are not usable in the future due to planned changes. 
 
@hyi This change fully implements the immutable flag. All routines report effective privilege. 

Summary of changes
---------------------------

1. Implemented necessary new routines for groups. 

   a. GroupAccess.view_users
   b. GroupAccess.edit_users
   c. GroupAccess.owners 

2. Standardized naming conventions for access control calls. 

Before | After | Comments 
-------- | ------- | ---------------
get_holding_users, get_holders, get_all_holders | view_users | Implemented consistently for resources and groups. 
get_held_\* | get_view_\* | same for resources and groups. 
get_editable_\* | get_edit_\* | same for resources and groups. 

The point of all of this is that:

Desired action | get users | get groups and resources
-------------- | ------------------- | -------------------------------
VIEW | r.view_users, g.view_users| u.get_view_groups(), u.get_view_resources(), g.get_view_resources()
CHANGE | r.edit_users, g.edit_users | u.get_edit_groups(), u.get_edit_resources(), g.get_edit_resources() 
OWNER | r.owners, g.owners | u.get_owned_groups(), u.get_owned_resources()

a. If you want something viewable, then you use get_view_\* or view_\* (depending upon whether you want  things to view or users). 
b. If you want something editable, then you use get_edit_\* or edit_\* (depemding upon whether you  want things or users) 
c. If you want something owned, then you use get_owned_\* or owners (depending upon whether you want things or users). 

What this version does not do: 
------------------------------

This is not the big change in semantics that is promised. The semantics of ownership remain the same. The methods, however, have their final names for that change. In other words, multiple people can still assign CHANGE to one object.  This will be fixed in the next PR. 

Also, I did not tackle the other major component of fixing semantics, which is to remove the undo_\* routines and replace with unshare_\* function.

pkdash commented:
@alvacouch I plan to do a final review of your latest changes today.

alvacouch commented:
let me make a general comment on naming. I need a bit of advice:  

- The best practice for methods that do nontrivial queries is to list them as get_X() rather than properties X. 
- In our system, we have implemented ResourceAccess.owners, edit_users, view_users, edit_groups, view_groups as properties to trick mezzanine into using them. This is rather obsolete now for most cases but it's still in the code base. This is not considered best practice. It's expedient. 
- Everything else is implemented under the get_X() pattern, including 
    - UserAccess.get_edit_resources(), get_view_resources(), get_owned_resources()
    - GroupAccess.get_edit_resources(), get_view_resources()
 
That being said, *I am unwilling to rename* ResourceAccess.owners, edit_users, view_users, edit_groups, or view_groups to best practice conventions (because of an *overwhelming number of references*). 

The basic question is what to do with UserAccess.\* and GroupAccess.\*. It seems to me that the expedient thing is to name everything according to conventions established -- for better or worse -- to handle resources. Then we have entirely consistent naming, to wit: 

- ResourceAccess.owners, edit_users, view_users
- UserAccess.owned_resources, edit_resources, view_resources. 
- GroupAccess.edit_resources, view_resources. 

I am willing to do this, even though it is not best practice, because *it is easier to remember.* Do you agree?

pkdash commented:
@alvacouch I think having a consistent naming pattern (even though it may not fit to the best practice rules) would make it easier to remember and use. So I would favor these changes. Hope it won't take too much of your time to make these changes. I know this PR has been out there for quite sometime. We should close this ASAP. Thanks.

alvacouch commented:
It is not a big deal. I don't use pycharm to do it. I use global replace in VIM. This allows me to substitute arbitrary patterns and to refactor get_x() to x very quickly. It literally takes longer to validate via global testing than to make the change.

alvacouch commented:
(I think it is worth locking down the naming *now*, in this PR, so that everyone will know it and I can write documentation around it.)

alvacouch commented:
@pkdash Made naming changes and documented in README.md for future reference.  IMHO ready to commit.

alvacouch commented:
@pkdash changes now documented in master document: 
https://docs.google.com/document/d/1c-3GXA_ALxr9tCTAX0oxOZQrbaB4KNWrM54aZxhflww/edit#

pkdash commented:
@alvacouch I am giving + 1 for this huge PR based on the fact that you have extensive set of unit tests for access_control and they all are passing. Due to the complex nature of the some of the database queries as result of the complex access rules that we have, it is very hard to do a proper review of such large number of changes in this PR. However, I believe through the code review process we got the code base for this app in a better position. Thanks for your patience as it took me a while to finish this review.

alvacouch commented:
@pkdash Many, many thanks for *your* patience. Now I can go on to create the next PR, which is a lot simpler!
